fairseq/modules/moe_layer.py

```diff
@@ -0,0 +1,464 @@
+        )
+        self.expert_id = distributed_utils.get_data_parallel_rank()
+        self.shuffle = args.base_shuffle
+        self.bloss_type = args.moe_bloss_type
+        self.bloss_weight = args.moe_bloss_weight
+
+        # Add a special attribute to the expert parameters, so we know not to sync their gradients
+        # 
+        for param in self.expert_network.parameters():
+            param.expert = True
+
+    def forward(self, input_features, *args, **kwargs):
+        features = input_features.reshape(-1, input_features.size(-1))
+        is_training = input_features.requires_grad
+
+        if self.shuffle and is_training:
+            # Send each token to a random worker, to break correlations within the batch
+            shuffle_sort = torch.randperm(features.size(0), device=features.device)
+            features = All2All.apply(features[shuffle_sort])
+
+        with torch.no_grad():
+            # Compute similarity of each token to each expert, for routing
+            token_expert_affinities = features.matmul(self.expert_centroids.transpose(0, 1))
+            # TODO @margsli should be identical to below, is the einsum faster?
+            # check that features and expert_centroids' dimensions are ordered correctly
+            # token_expert_affinities = torch.einsum('sbh,eh->sbe', features, self.expert_centroids)
+            # TODO @margsli this is only argmax, should the other sampling strategies be implemented?
+            self.last_p = torch.softmax(token_expert_affinities, -1)
+
+        # Compute which token goes to which expert
+        routing_strategy = 'greedy'
+        sort_by_expert, input_splits, output_splits = self.assignment(token_expert_affinities, strategy=routing_strategy)
+        # Swap these tokens for the right ones for our expert
+        routed_features = All2All.apply(features[sort_by_expert], output_splits, input_splits)
+
+        if routed_features.size(0) > 0:
+            # Mix in the expert network based on how appropriate it is for these tokens
+            alpha = torch.sigmoid(routed_features.mv(self.expert_centroids[self.expert_id])).unsqueeze(1)
+            routed_features = alpha * self.expert_network(routed_features) + (1 - alpha) * routed_features
+        # Return to original worker and ordering
+        result = All2All.apply(routed_features, input_splits, output_splits)[self.inverse_sort(sort_by_expert)]
+
+        if self.shuffle and is_training:
+            # Undo shuffling
+            result = All2All.apply(result)[self.inverse_sort(shuffle_sort)]
+
+        # Return additional Nones for compatibility with TransformerDecoderLayer
+        return result.view(input_features.size()), None, None
+
+    def inverse_sort(self, order):
+        # Creates an index that undoes a sort: xs==xs[order][inverse_sort(order)]
+        return torch.empty_like(order).scatter_(0, order, torch.arange(0, order.size(0), device=order.device))
```

Comment:
It is easy to introduce an error here and the network still works. I find it difficult to check it mentally. Maybe we can talk about this in the meeting and run through the dimensions etc